### PR TITLE
chore: add name validation on db/app creation

### PR DESCRIPTION
**Pull request by Akirtovskis**
can be merged only after : https://github.com/ledokku/ledokku/pull/172

@pradel 
- this pr adds name validation for both creating db and creating app 
- for the app db it should be 💯 good to go
- for db creation I am struggling with the validation for `one name per db type` as I imagine we wouldn't want to limit name `to one db name per server` so I added a comment below to point out the problematic part

### CODE REVIEW

File: client/src/pages/create-database.tsx [hunk 1]
```diff
@@ -63,10 +64,67 @@ const DatabaseBox = ({ label, selected, icon, onClick }: DatabaseBoxProps) => {
...
+      .string()
+      .oneOf(['POSTGRESQL', 'MYSQL', 'MONGODB', 'REDIS'])
+      .required(),
+    name: yup
+      .string()
+      .required('Database name is required')
+      .matches(/^[a-z0-9-]+$/)
+      .when(['type'], {
+        is: (val) => val === 'POSTGRESQL',
+        // val === databaseQueryData?.databases.find((db) => db.type === val),
```
[the remark this answers is not in the document]
Comment (pradel) [reply client/src/pages/create-database.tsx:111]: I guess the best to check if the name is free is to actually make a new request to the server every time the user changes the name (with some denounce of course).

File: client/src/pages/create-app.tsx [hunk 2]
```diff
@@ -48,27 +69,37 @@ export const CreateApp = () => {
...
-            >
-              Create
-            </Button>
+        <form>
+          <div className="grid grid-cols-1 md:grid-cols-3">
+            <div>
+              <FormLabel>App name: </FormLabel>
+              <FormInput
+                autoComplete="off"
+                // className="inline w-full max-w-xs bg-white border border-grey rounded py-3 px-3 text-sm leading-tight transition duration-200 focus:outline-none focus:border-black"
```
Comment (pradel) [inline client/src/pages/create-app.tsx:78]: @Akirtovskis let's remove this before merging?

File: client/src/pages/create-app.tsx [hunk 3]
```diff
@@ -1,18 +1,39 @@
...
+
+  const createAppSchema = yup.object().shape({
+    name: yup
+      .string()
+      .required('App name is required')
+      .matches(/^[a-z0-9-]+$/)
+      .test(
+        'Name exists',
+        'App with this name already exists',
+        (val) => !data?.apps.find((app) => app.name === val)
```
Comment (pradel) [inline client/src/pages/create-app.tsx:27]: @Akirtovskis we should also check the type of the db with the name I think

File: client/src/pages/create-app.tsx [hunk 4]
```diff
@@ -1,18 +1,39 @@
...
   const formik = useFormik<{ name: string }>({
     initialValues: {
       name: '',
     },
+
+    validateOnChange: true,
+    validationSchema: createAppSchema,
     // TODO yup validation
     onSubmit: async (values) => {
       // TODO validate name
```
Comment (pradel) [inline client/src/pages/create-app.tsx:39]: you can remove this TODO

File: client/src/pages/create-database.tsx [hunk 5]
```diff
@@ -203,7 +222,11 @@ export const CreateDatabase = () => {
               ) : null}
             </React.Fragment>
           ) : (
-            <form onSubmit={formik.handleSubmit} className="mt-8">
+            <form
+              onSubmit={formik.handleSubmit}
+              onChange={formik.handleChange}
```
Comment (pradel) [inline client/src/pages/create-database.tsx:227]: Hum I am not sure about this onChange 🤔 I never saw that before

File: server/src/graphql/mutations/createApp.ts [hunk 6]
```diff
@@ -14,6 +14,14 @@ export const createApp: MutationResolvers['createApp'] = async (
     throw new Error('Unauthorized');
   }
 
+  const apps = await prisma.app.findMany();
```
Comment (pradel) [inline server/src/graphql/mutations/createApp.ts:17]: would be better to do a `findMany` with the name directly and check the item 0 in the array if it's set

Comment (pradel) [reply server/src/graphql/mutations/createApp.ts:17]: This step should be done after the yup validation :)

File: server/src/graphql/mutations/createDatabase.ts [hunk 7]
```diff
@@ -25,6 +25,15 @@ export const createDatabase: MutationResolvers['createDatabase'] = async (
     throw new Error('Unauthorized');
   }
 
+  const databases = await prisma.database.findMany();
```
Comment (pradel) [inline server/src/graphql/mutations/createDatabase.ts:28]: same comments as server/src/graphql/mutations/createApp.ts :)

File: server/src/graphql/mutations/createDatabase.ts [hunk 8]
```diff
@@ -28,6 +28,18 @@ export const createDatabase: MutationResolvers['createDatabase'] = async (
   // We make sure the name is valid to avoid security risks
   createDatabaseSchema.validateSync({ name: input.name });
 
+  const databases = await prisma.database.findMany({
+    where: {
+      name: input.name,
```
Comment (pradel) [inline server/src/graphql/mutations/createDatabase.ts:33]: @Akirtovskis you can also add the type in the where condition that way you don't need the find afterwards :)

Comment (Akirtovskis) [reply server/src/graphql/mutations/createDatabase.ts:33]: @pradel Done! 🙌

**Review by pradel — APPROVED** [review]
🚀

**Pull request merged**